bigquery/google/cloud/bigquery/_pandas_helpers.py

```diff
@@ -176,5 +207,163 @@ def to_parquet(dataframe, bq_schema, filepath):
+    """Separate finished and unfinished threads, much like
+    :func:`concurrent.futures.wait`, but don't wait.
+    """
+    done = []
+    not_done = []
+    for future in futures:
+        if future.done():
+            done.append(future)
+        else:
+            not_done.append(future)
+    return done, not_done
+
+
+def download_dataframe_bqstorage(
+    project_id,
+    table,
+    bqstorage_client,
+    dtypes,
+    preserve_order=False,
+    selected_fields=None,
+):
+    """Use (faster, but billable) BQ Storage API to construct DataFrame."""
+    if "$" in table.table_id:
+        raise ValueError(
+            "Reading from a specific partition is not currently supported."
+        )
+    if "@" in table.table_id:
+        raise ValueError("Reading from a specific snapshot is not currently supported.")
+
+    read_options = bigquery_storage_v1beta1.types.TableReadOptions()
+    if selected_fields is not None:
+        for field in selected_fields:
+            read_options.selected_fields.append(field.name)
+
+    requested_streams = 0
+    if preserve_order:
+        requested_streams = 1
+
+    session = bqstorage_client.create_read_session(
+        table.to_bqstorage(),
+        "projects/{}".format(project_id),
+        read_options=read_options,
+        requested_streams=requested_streams,
+    )
+
+    # We need to parse the schema manually so that we can rearrange the
+    # columns.
+    schema = json.loads(session.avro_schema.schema)
+    column_names = [field["name"] for field in schema["fields"]]
+
+    # Avoid reading rows from an empty table. pandas.concat will fail on an
+    # empty list.
+    if not session.streams:
+        yield pandas.DataFrame(columns=column_names)
+        return
+
+    total_streams = len(session.streams)
+
+    # Use _DownloadState to notify worker threads when to quit.
+    # See: https://stackoverflow.com/a/29237343/101923
+    download_state = _DownloadState()
+
+    # Create a queue to collect frames as they are created in each thread.
+    worker_queue = queue.Queue()
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=total_streams) as pool:
+        try:
+            # Manually submit jobs and wait for download to complete rather
+            # than using pool.map because pool.map continues running in the
+            # background even if there is an exception on the main thread.
+            # See: https://github.com/googleapis/google-cloud-python/pull/7698
+            not_done = [
+                pool.submit(
+                    _download_dataframe_bqstorage_stream,
+                    download_state,
+                    bqstorage_client,
+                    column_names,
+                    dtypes,
+                    session,
+                    stream,
+                    worker_queue,
+                )
+                for stream in session.streams
+            ]
+
+            while not_done:
+                done, not_done = _nowait(not_done)
```

Comment:
There are two things that I'm waiting for: the worker queue (all the DataFrames as they are processed from each thread) and for the threads to complete. I want the worker queue to take precedence.

If the queue fills up, it can block processing because the threads will block until an item is pulled from the queue. I actually encountered this behavior when I first implemented this code, as it manifested as much slower than expected download speeds. See the following picture:

![wait for queue vs  wait for futures (1)](https://user-images.githubusercontent.com/247555/59523755-f9401880-8e86-11e9-9c88-06f0886b0769.png)

Notice that most of the `put` calls block because of a full queue.

One way to address the full queue is to have an inner loop to drain the queue each time. I actually tried this, too, but it resulted in bursty reads. Many `put` calls still would block.

![wait for queue vs  wait for futures (2)](https://user-images.githubusercontent.com/247555/59523840-2ee50180-8e87-11e9-9643-4bc40d766ccd.png)

What I've implemented instead is to not ever wait for the threads to finish, and instead block on the queue.

![wait for queue vs  wait for futures](https://user-images.githubusercontent.com/247555/59523872-4623ef00-8e87-11e9-995b-40e967f0fbfe.png)

This could result in waiting longer than we need to at the end, since we could wait for a full `_PROGRESS_INTERVAL` before noticing that all streams are finished. Since these streams are expected to last quite long compared to the `_PROGRESS_INTERVAL`, I'm not as concerned about that. My main motivation is to ensure that `put` calls do not block.

Reply:
Great, thanks, makes perfect sense!